walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,174 @@
+   * chosen[]: temporary array to store current combination.
+   * start & end: staring and ending indexes in arr[].
+   * index: current index in data[].
+   */
+  private void combinationUtil(String arr[], String chosen[], int start, int end, int index, int r) { 
+    // If current combination is ready to be added, then add it.
+    if (index == r) {
+      optionalAttendeesCombs.add(Arrays.asList(chosen));
+      return;
+    }
+
+    // Replace index with all possible elements. 
+    // The condition "end-i+1 >= r-index" makes sure that including one element
+    // at index will make a combination with remaining elements at remaining positions.
+    for (int i = start; i <= end && end - i + 1 >= r - index; i++) { 
+      chosen[index] = arr[i];
+      combinationUtil(arr, chosen, i + 1, end, index + 1, r);
+    }
+  }
+
+  /**
+   * Returns all combinations of size r in arr[], which has size n.
+   * This function mainly uses combinationUtil().
+   */
+  private void getAllCombinations(String arr[], int n, int r) { 
+    optionalAttendeesCombs = new ArrayList<>();
+
+    // A temporary array to store all combination one by one.
+    String chosen[] = new String[r]; 
+
+    combinationUtil(arr, chosen, 0, n - 1, 0, r); 
+  }
+
+  /** 
+   * Finds a collection of timeRanges that are gaps among the timeRanges in {@code eventTimes}
+   * that have lengths longer than {@code duration}.
+   */
+  private Collection<TimeRange> findTimeRangeGaps(List<TimeRange> eventTimes, long duration) {
+    Collection<TimeRange> possibleTimes = new ArrayList<>();
+    int lastEndTime = TimeRange.START_OF_DAY;
+    TimeRange currentRange;
+    for (int i = 0; i < eventTimes.size(); i++) {
+      currentRange = eventTimes.get(i);
+      if (currentRange.start() <= lastEndTime) {
+        // If there is an overlap, update the previous end marker.
+        lastEndTime = Math.max(lastEndTime, currentRange.end());  
+      } else {
+        // If there is a gap, check if it can fit the requested event.
+        if (currentRange.start() - lastEndTime >= duration) {
+          possibleTimes.add(TimeRange.fromStartEnd(lastEndTime, currentRange.start(), false));
+        }
+        lastEndTime = currentRange.end();
+      }
+    }
+
+    if (TimeRange.END_OF_DAY - lastEndTime >= duration) {
+      possibleTimes.add(TimeRange.fromStartEnd(lastEndTime, TimeRange.END_OF_DAY, true));
+    }
+    return possibleTimes;
+  }
+
+  /**
+   * Loads {@code attendeesEventTimes} with the time ranges of events that involve any of the mandatory attendees.
+   * Loads {@code optionalAttendeesEventTimes} with a map from the names of optional attendees 
+   * to lists of time ranges of events that involve them.
+   */
+  private void getAttendeesEventTimes(
+      Collection<Event> events, Collection<String> attendees, Collection<String> optionalAttendees) {
+    attendeesEventTimes = new ArrayList<>();
+    optionalAttendeesEventTimes = new HashMap<>();
```

Comment:
For class attributes you want to use `this.attributeName`

Reply:
got it